google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java

```diff
@@ -0,0 +1,247 @@
+    // The cache of page tokens. It is stored in-memory and in ascendant order of page number.
+    // The first page's token is always an empty string.
+    List<String> pageTokens = new ArrayList<>(Collections.singletonList(""));
+
+    System.out.println("--- 0. Fetch page 1 to get metadata:");
+
+    // Creates a query that retrieves the campaigns.
+    String query =
+        String.format(
+            "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT %d",
+            RESULTS_LIMIT);
+
+    // Creates a paginated search request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(Long.toString(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .setReturnTotalResultsCount(true)
+            .build();
+
+    int totalNumberOfPages;
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Issues a paginated search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      cacheNextPageToken(pageTokens, response.getPage(), 1);
+
+      // Determines the total number of results and prints it.
+      // The total results count does not take into consideration the LIMIT clause of the query,
+      // so we need to find the minimal value between the limit and the total results count.
+      long totalNumberOfResults =
+          Math.min(RESULTS_LIMIT, response.getPage().getResponse().getTotalResultsCount());
+      System.out.printf("Total number of campaigns found: %d.%n", totalNumberOfResults);
+
+      // Determines the total number of pages and prints it.
+      totalNumberOfPages = (int) Math.ceil((double) totalNumberOfResults / (double) PAGE_SIZE);
+      System.out.printf("Total number of pages: %d.%n", totalNumberOfPages);
+      if (totalNumberOfPages == 0) {
+        System.out.println("Could not find any campaigns.");
+        System.exit(1);
+      }
+    }
+
+    // Demonstrates how the logic works when iterating pages forward. We select a page that is
+    // in the middle of the result set so that only a subset of the page tokens will be cached.
+    int middlePageNumber = (int) Math.ceil((double) totalNumberOfPages / 2.0);
+    System.out.printf("--- 1. Print results of the middle page (page %d):%n", middlePageNumber);
+    fetchAndPrintPageResults(googleAdsClient, customerId, query, middlePageNumber, pageTokens);
+
+    // Demonstrates how the logic works when iterating pages backward with some page tokens that
+    // are not already cached.
+    System.out.println("--- 2. Print results from the last page to the first:");
+    for (int pageNumber = totalNumberOfPages; pageNumber > 0; pageNumber--) {
+      System.out.printf("-- Printing results for page %d:%n", pageNumber);
+      fetchAndPrintPageResults(googleAdsClient, customerId, query, pageNumber, pageTokens);
+    }
+  }
+
```

Comment:
In terms of the output a user will see on the console, I think just "Results of the xxx:" is better. "Print results..." reads like code comments.